packages/shared/src/components/squads/SquadFeedHeading.tsx

```diff
@@ -0,0 +1,65 @@
+import React, { ReactElement, useContext, useMemo } from 'react';
+import { Button } from '../buttons/Button';
+import { PinIcon } from '../icons';
+import { Squad } from '../../graphql/sources';
+import { ActiveFeedContext } from '../../contexts';
+import { useSquadActions } from '../../hooks';
+import { LoggedUser } from '../../lib/user';
+
+interface SquadFeedHeadingProps {
+  squad: Squad;
+  user: LoggedUser;
+}
+
+function SquadFeedHeading({
+  squad,
+  user,
+}: SquadFeedHeadingProps): ReactElement {
+  const { items } = useContext(ActiveFeedContext);
+  const { collapseSquadPinnedPosts, expandSquadPinnedPosts } = useSquadActions({
+    squad,
+    user,
+  });
+  const collapsePinnedPosts = squad?.currentMember?.flags?.collapsePinnedPosts;
+  const itemsCount = items.length;
+
+  const onClick = async () => {
+    return collapsePinnedPosts
+      ? await expandSquadPinnedPosts(squad.id)
+      : await collapseSquadPinnedPosts(squad.id);
+  };
+
+  const pinnedPostsCount = useMemo(() => {
+    let countPinned = 0;
+    // eslint-disable-next-line no-plusplus
+    for (let i = 0; i < items.length; i++) {
+      const item = items[i];
+      if (item.type === 'post' && !!item.post.pinnedAt) {
+        countPinned += 1;
+      } else {
+        break;
+      }
+    }
+
+    return countPinned;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [itemsCount]);
```

Comment:
Maybe we should use reduce here, I recall it being a little bit faster, and also a bit cleaner.

Reply:
`break` here will terminate the loop early if there is a post that is not pinned in array, so the count will be wrong

agree `reduce` will be cleaner here

also, why do we need `eslint-disable` here?

Reply:
the goal here was to have the counter break early from the iteration since we know pinned posts will always be at the beginning of the list, so it would be inefficient to iterate through the whole array when we will know the pinned posts count after the first few iterations.

I agree that it does not look clean (although it's a little bit funny how vanilla code seems `dirty` now to us 😅 ), but if we are fine with the performance hit here, we can change it to a `reduce` or something else.

> also, why do we need eslint-disable here?

@capJavert I'm using the `items.length` as the dependency which causes linter to complain because it is expecting the `items` object to be passed in

Reply:
> the goal here was to have the counter break early from the iteration since we know pinned posts will always be at the beginning of the list, so it would be inefficient to iterate through the whole array when we will know the pinned posts count after the first few iterations.

Oh I see, I think problem here with this optimization is that it takes for granted that pinned posts will be first, if that was to change the code would not work. I think there is no need to do a premature optimization here.

> @capJavert I'm using the items.length as the dependency which causes linter to complain because it is expecting the items object to be passed in

If squad feed refetches and we get same number of items but different posts are pinned then the count would again be wrong. This is the exact case where that lint rule helps to avoid edge cases and unintentional bugs down the line.

Reply:
updated 🙏  I still feel like we are sacrificing performance for readability here, but I guess since the squads are paginated we will only ever go through a few 10s of records when doing this, so that's fine